### PR TITLE
HistogramData rollout: algorithms LoadQKK to LoadSpiceXML2DDet

### LINKED ISSUE #18495: HistogramData rollout: algorithms LoadQKK to LoadSpiceXML2DDet
This issue is part of the HistogramData rollout, see https://github.com/mantidproject/mantid/issues/17641.
Algorithms covered in this issue are:

```
Framework/DataHandling/src/LoadQKK.cpp
Framework/DataHandling/src/LoadRawHelper.cpp
Framework/DataHandling/src/LoadRKH.cpp
Framework/DataHandling/src/LoadSassena.cpp
Framework/DataHandling/src/LoadSINQFocus.cpp
Framework/DataHandling/src/LoadSNSspec.cpp
Framework/DataHandling/src/LoadSpec.cpp
Framework/DataHandling/src/LoadSPE.cpp
Framework/DataHandling/src/LoadSpice2D.cpp
Framework/DataHandling/src/LoadSpiceXML2DDet.cpp
```

### PR DESCRIPTION
**Pull request by samueljackson92**
## Description of work.
This PR is part of refactoring effort to use the new interface of the HistogramData module.
See https://github.com/mantidproject/mantid/issues/17641 for details.

The following algorithms were refactored to use HistogramData:

Performance tests were added for: N/a

The following algorithms have improved performance: N/a

## To test.

Code review.

Fixes #18495.
Internal change, no release notes (potential performance improvements will be added to release notes as part of the umbrella issue).

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataHandling/src/LoadSINQFocus.cpp [hunk 1]
```diff
@@ -161,21 +161,23 @@ void LoadSINQFocus::loadDataIntoTheWorkSpace(NeXus::NXEntry &entry) {
...
+  x.assign(timeBinning.begin(), timeBinning.end());
 
   Progress progress(this, 0, 1, m_numberOfTubes * m_numberOfPixelsPerTube);
   size_t spec = 0;
   for (size_t i = 0; i < m_numberOfTubes; ++i) {
     for (size_t j = 0; j < m_numberOfPixelsPerTube; ++j) {
       if (spec > 0) {
         // just copy the time binning axis to every spectra
-        m_localWorkspace->dataX(spec) = m_localWorkspace->readX(0);
+        m_localWorkspace->mutableX(spec) = x;
```
Comment (mantid-roman) [inline Framework/DataHandling/src/LoadSINQFocus.cpp]: Can `setShared` be used here?

File: Framework/DataHandling/src/LoadSpec.cpp [hunk 2]
```diff
@@ -50,109 +50,132 @@ void LoadSpec::init() {
...
+ * Convert the data from the input buffer to a histogram object
+ *
+ * @param input :: the input buffer containing the raw data
+ * @param histogram :: the histogram object to fill with values
+ * */
+void LoadSpec::readHistogram(const std::vector<double> &input,
+                             HistogramData::Histogram &histogram) const {
+  std::vector<double> x, y, e;
+
+  size_t index = 0;
```
Comment (mantid-roman) [inline Framework/DataHandling/src/LoadSpec.cpp]: Why is `index` defined outside of `for`?

File: Framework/DataHandling/src/LoadSpec.cpp [hunk 3]
```diff
@@ -50,109 +50,132 @@ void LoadSpec::init() {
...
+  for (; index < input.size() - 1; index++) {
```
Comment (mantid-roman) [inline Framework/DataHandling/src/LoadSpec.cpp]: Is the size of `input` guaranteed to make the loop valid?

File: Framework/DataHandling/src/LoadSINQFocus.cpp (hunk 1 above)
Comment (samueljackson92) [inline Framework/DataHandling/src/LoadSINQFocus.cpp]: Yes I think so.

File: Framework/DataHandling/src/LoadSpec.cpp (hunk 2 above)
Comment (samueljackson92) [inline Framework/DataHandling/src/LoadSpec.cpp]: Hang over from refactoring. Agreed, this should be changed.

File: Framework/DataHandling/src/LoadSpec.cpp (hunk 3 above)
Comment (samueljackson92) [inline Framework/DataHandling/src/LoadSpec.cpp]: I think this was just based on how it worked at line 121 in the original file. But you're right we should probably not make assumptions when we can calculate a proper upper limit from the size of `input`.

**Comment by mantid-roman** [comment]
Looks good. :shipit: The failed test seems unrelated.

**Pull request merged**